### DISCUSSION TITLE
Data splitting and dataset.data

### DISCUSSION THREAD

**Discussion by anthonysirico** [pyg-team/pytorch_geometric#5607] [Q&A] [answered]
Is there a better way to split the dataset besides doing `training = dataset[:1000]` `val = dataset[1000:1500]` etc? Because I ran into two issues doing it this way.

1) I want to be able to sort the subset, is there a way to do that? I have `dataset.shuffle()` within the code, but is there a way to do it without re-writing the InMemoryDataset class, i.e. from the console?
2) Why is `dataset.data `the same as `training.data` and `val.data`? Why would a subset of the large dataset still show `.data` of the large dataset and not its own?

**Comment by rusty1s** [comment, marked as answer]
1. Yes, you can sort based on a pre-defined permutation, e.g.:
```python
dataset = dataset[torch.tensor([0, 2, 4, 6])]
```
2. Shuffling/Splitting does not modify the underlying data (the same as a tensor slice will not copy the tensor). This allows us to not modify/copy the underlying data. If you strictly want to enforce this, try `dataset = dataset[:1000].copy()`.

**Reply by anthonysirico** [reply]
For 1., how does that help if I want to sort the dataset by `dataset.data.y` for example? 

What I am doing is creating the dataset without the `y` feature. then I am splitting the dataset for the purposes of my experiment, and for one of the subsets, I want to sort it based on a dataset feature.
```
dataset = CircuitDataset('test')
dataset.shuffle()
split = 0.5
N = round(len(dataset)*split)
known = dataset[:N].copy()
unknown = dataset[N:].copy()
```
At this point, I want to sort the known set based on `known.data.performance`, then assign its `y`  graph label.

**Reply by anthonysirico** [reply]
also, it seems that after creating the dataset, I can't do the following:
`dataset[0].y = torch.tensor(np.asarray([1]), dtype=torch.int64)`

So can I keep y as a separate list of tensors since I am changing y for each experiment?

**Reply by rusty1s** [reply]
1. If you want to sort by `y`, you can do something like
```python
perm = dataset.data.y.argsort()
dataset = dataset[perm].copy()
```
Nonetheless, please note that modifying `dataset.data` is not recommened. If you want to make modifications to your data, you should do this via the concept of transforms:
```python
def my_transform(data):
    data.y = torch.tensor(...)
    return data

dataset = MyDataset(..., transform=my_transform)
```

**Reply by anthonysirico** [reply]
I think i am going to do that way in my other question, just splitting everything and have it return two datasets ready to go.

**Reply by anthonysirico** [reply]
@rusty1s I am still stuck on the sorting issue. If my dataset has an attribute `dataset.data.performance`, how can I sort the set based on performance? My previous attempt didn't work out.

**Reply by anthonysirico** [reply]
the sort worked, thanks.

**Reply by anthonysirico** [reply]
@rusty1s whats wrong with this code?

```
def add_graph_label(data):
    for i in range(0,10812):
        if i <= 10812*0.25:
            binary = np.asarray([1])
            data.y = torch.tensor(binary, dtype=torch.int64)
        else:
            binary = np.asarray([0])
            data.y = torch.tensor(binary, dtype=torch.int64)
    return data

known_training = ImbCircuitDataset('test_imb', split=0.25, transform=add_graph_label)
```
when I look up `known_training[0].y`, they are all `0`. Also, when I run `known_training.data`, there is no `y`.

**Reply by rusty1s** [reply]
I am not sure what's the usage of your for-loop. `add_graph_label` will get called for each `data` object individually. If you want to perform transforms based on indices. It may be best to do
```python
class MyTransform:
    def __init__(self):
         self.i = 0

    def __call__(self, data):
        if self.i <= 10812*0.25:
            binary = np.asarray([1])
            data.y = torch.tensor(binary, dtype=torch.int64)
        else:
            binary = np.asarray([0])
            data.y = torch.tensor(binary, dtype=torch.int64)
        self.i += 1
         return data

known_training = ImbCircuitDataset('test_imb', split=0.25, pre_transform=MyTransform())
```
(note the usage of `pre_transform` to only apply the transformation once).

**Reply by anthonysirico** [reply]
i tried it and now i have no way of checking. 
```
known_training[0]
```
returns
```
TypeError                                 Traceback (most recent call last)
Input In [12], in <cell line: 1>()
----> 1 known_training[0]

File ~\anaconda3\lib\site-packages\torch_geometric\data\dataset.py:198, in Dataset.__getitem__(self, idx)
    193 if (isinstance(idx, (int, np.integer))
    194         or (isinstance(idx, Tensor) and idx.dim() == 0)
    195         or (isinstance(idx, np.ndarray) and np.isscalar(idx))):
    197     data = self.get(self.indices()[idx])
--> 198     data = data if self.transform is None else self.transform(data)
    199     return data
    201 else:

TypeError: __init__() takes 1 positional argument but 2 were given
```
My code:
```
class _add_graph_labels:
    def __init__(self):
        self.i = 0

    def __call__(self, data):
        if self.i <= 10812*0.25:
            binary = np.asarray([1])
            data.y = torch.tensor(binary, dtype=torch.int64)
        else:
            binary = np.asarray([0])
            data.y = torch.tensor(binary, dtype=torch.int64)
        self.i += 1
        return data
```
I originally wrote it like yours, but you didn't have `self.i += 1` and the `return` indented and my IDE was throwing an error.

If this doesn't work, could I create a separate list for y using a loop?
```
y = []
for i in range(len(known_sorted)):
    if i <= len(known_sorted)*0.25:
        binary = np.asarray([1])
        y.append(torch.tensor(binary, dtype=torch.int64))
    else:
        binary = np.asarray([0])
        y.append(torch.tensor(binary, dtype=torch.int64))
```
I just don't know how that would be used in combination with the DataLoader.

**Reply by rusty1s** [reply]
Keep in mind that you need to pass the transform as `transform=MyTransform()`. I fixed my above code snippet.
